snap/quota/quota.go

```diff
@@ -384,12 +429,23 @@ func (grp *Group) validateMemoryResourceFit(allQuotas map[string]*groupQuotaAllo
 // limit of the group, and if not locates the nearest parent group that has a cpu quota, and then verifies
 // if that group has any space available by checking its 'cpuReserved'. The 'cpuReserved' tells us how much
 // of the group quotas limit has been used already by its subgroups (excluding the one querying).
-func (grp *Group) validateCPUResourceFit(allQuotas map[string]*groupQuotaAllocations, cpuCount, cpuPercentage int) error {
+func (grp *Group) validateCPUResourceFit(allQuotas map[string]*groupQuotaAllocations, resourceLimits Resources) error {
+
+	// calculate the requested cpu usage in the context of this group, we need
+	// to first of all calculate the final cpu usage from the provided cpu set, and
+	// any existing limitations.
+	cpuCountRequested := grp.GetCorrectedCPUCount()
+	if resourceLimits.CPU.Count != 0 {
+		cpuCountRequested = resourceLimits.CPU.Count
+	}
+	if resourceLimits.CPUSet != nil && len(resourceLimits.CPUSet.CPUs) > 0 {
+		cpuCountRequested = len(resourceLimits.CPUSet.CPUs)
```

Comment:
Shouldn't you consider the minimum value between the `resourceLimits.CPU.count` and `len(resourceLimits.CPUSet.CPUs)`? That is, doing something like this:

```suggestion
	if resourceLimits.CPUSet != nil {
		cpuSetLength = len(resourceLimits.CPUSet.CPUs)
		if cpuSetLength > 0 && cpuSetLength < cpuCountRequested{
			cpuCountRequested = len(resourceLimits.CPUSet.CPUs)
		}
```

Reply:
Yea to be honest you are first of all right, and secondly I have some weaknesses in my approach to CPU.Count which I need to revise

Reply:
@mardy I have revised this code, and I want to explain here why the code still doesn't take `len(resourceLimits.CPUSet.Set)` into account when using the `.count`, the `.count` should be viewed as a multiplier that gets applied to the percentage. If the multiplier is set, it will be used, otherwise the multiplier will always default to the number of allowed cores.

`--cpu=8x10%` is a valid option for the cpu quota, and this sets the `count` to 8
`--cpu=80%` needs to calculate the count based on circumstances.

This is the purpose I'm trying to convey, in this small refactor I just did I hope I have clarified this enough in the docs.